jenkinsci/tfs-plugin
# The TFVC SCM can be used with collection credential pairs

Changes
-------

1. **Trivial:** Improved the wording of the help text in the global configuration:
![image](https://cloud.githubusercontent.com/assets/297515/17820664/e399292c-661b-11e6-98f2-6aa9dd78bdf2.png)
2. **Minor:** The _Collection URL_ field is now a combobox populated from the values entered in the global configuration (i.e. the above interface), while still allowing values to be typed manually:
![image](https://cloud.githubusercontent.com/assets/297515/17820730/3b8ba79a-661c-11e6-93bc-978bbde5e356.png)
3. **Major:** Added a _Credentials_ field that supports two values: **Automatic** and **Manual**.
    1. New jobs default to **Automatic** and will use values configured in the global configuration (as documented in the help):
![image](https://cloud.githubusercontent.com/assets/297515/17820868/d8ba2b18-661c-11e6-83a9-1ae35b26c68b.png)
    2. Existing jobs with manually-specified credentials will take the value of **Manual** to avoid disruption:
![image](https://cloud.githubusercontent.com/assets/297515/17820967/4d8526f0-661d-11e6-8308-ade334dd053c.png)


Manual testing
--------------

1.  **Upgrade**
    1. Using a previous version of the TFS plugin, create a job that uses the TFVC SCM.
        1. I used a Team Services account.
    2. Upgrade the plugin to the version containing these changes.
    3. Queue a build of the job.
        1. It ran successfully.
    4. **Configure** the job.
        1. The _Credentials_ field is set to **Manual** and the credentials are still there.
    5. Click **Save** or **Apply** without changing anything.
        1. Inspecting the job's `config.xml` file, one still sees entries for `userName` and `password` (they haven't changed), as well as a new line that reads `<credentialsConfigurer class="hudson.plugins.tfs.model.ManualCredentialsConfigurer"/>`.
    6. Queue another build of the job.
        1. It ran successfully.
2. **Migrate**
    1. Follow the instructions in the **Global configuration** section of the README.md to add _Credentials_ and a _Team Project Collection_ that uses those credentials.
        1. I used _alternate credentials_ that matched those used in the above job.
    2. **Configure** the upgraded job.
    3. Change its _Credentials_ field to **Automatic**.
    4. Click **Save** or **Apply**.
        1. Inspecting the job's `config.xml` file, the entries for `userName` and `password` were removed and the `credentialsConfigurer` line now read: `<credentialsConfigurer class="hudson.plugins.tfs.model.AutomaticCredentialsConfigurer"/>`.
    5. Queue another build of the job.
        1. It ran successfully.
3. **Change credentials**
    1. Find the _Credentials_ entry created in the previous scenario and select **Update**.
    2. Go create a _Personal Access Token_ (PAT) for the user account used earlier.
        1. I assigned the following scopes: `Code (read)`, `Code (status)` and `Work items (read and write)`.
    3. Copy-paste the new PAT in the **Password** field.
    4. Click **Save**.
    5. Queue another build of the job.
        1. It failed with `TFSUnauthorizedException: Access denied connecting to TFS server...` This is because Team Services does not yet support scoped PATs.
    6. Go edit the PAT and grant it `All scopes`.
    7. Queue another build of the job.
        1. It ran successfully.
4. **New job**
    1. **Jenkins** > **New Item**.
    2. Give the job a name, select **Freestyle project** and click **OK**.
    3. Select **TFVC** as the _SCM_.
        1. Notice the _Credentials_ field has the value of **Automatic**.
    4. Give focus to the _Collection URL_ field.
        1. A drop-down list appeared below the textbox with the URL to the team project collection added in the second scenario. (just like the 2nd screenshot)
    5. Select the highlighted entry.
    6. Enter a valid value for _Project path_.
    7. Add a _build step_.
        1. I added an _Execute shell_ with a _Command_ of `ls -l`.
    8. Click **Save**.
    9. Queue a build of the job.
        1. It ran successfully.
5. **Polling**
    1. Configured the job from the previous scenario to poll every minute.
    2. Clicked on  **Team Foundation Version Control (TFVC) Polling Log** and saw `No changes`.
    3. Generated some changesets in the TFVC repository.
        1. Four builds were triggered by activity in the TFVC repository.

Mission accomplished!
